bindings/rust/wasmedge-sys/src/vm.rs

```diff
@@ -0,0 +1,97 @@
+use super::wasmedge;
+use crate::{
+    raw_result::{decode_result, ErrReport},
+    string::StringRef,
+    value::Value,
+};
+
+// If there is a third-party sdk based on wasmedge-sys
+// the private encapsulation  here can force the third-party sdk to use the api
+#[derive(Debug)]
+pub struct Vm {
+    pub(crate) ctx: *mut wasmedge::WasmEdge_VMContext,
+    _private: (),
+}
+
+impl Vm {
+    pub fn create(config: &crate::config::Config) -> Result<Self, ErrReport> {
+        let ctx = unsafe {
+            wasmedge::WasmEdge_VMCreate(config.ctx, std::ptr::null_mut() /* store */)
+        };
+        if ctx.is_null() {
+            return Err(ErrReport::default());
+        }
+        Ok(Self { ctx, _private: () })
+    }
+
+    pub fn load_wasm_from_ast_module(
+        self,
+        module: &crate::module::Module,
+    ) -> Result<Self, ErrReport> {
+        unsafe {
+            decode_result(wasmedge::WasmEdge_VMLoadWasmFromASTModule(
+                self.ctx, module.ctx,
+            ))?;
+        }
+        Ok(self)
+    }
+
+    pub fn validate(self) -> Result<Self, ErrReport> {
+        unsafe {
+            decode_result(wasmedge::WasmEdge_VMValidate(self.ctx))?;
+        }
+        Ok(self)
+    }
+
+    pub fn instantiate(self) -> Result<Self, ErrReport> {
+        unsafe {
+            decode_result(wasmedge::WasmEdge_VMInstantiate(self.ctx))?;
+        }
+        Ok(self)
+    }
+
+    pub fn run(
+        &mut self,
+        func_name: impl AsRef<str>,
+        params: &[Value],
+    ) -> Result<Vec<Value>, ErrReport> {
+        // construct func
+        let raw_func_name: wasmedge::WasmEdge_String = StringRef::from(func_name.as_ref()).into();
+
+        let raw_params: Vec<_> = params
+            .as_ref()
+            .iter()
+            .copied()
+            .map(wasmedge::WasmEdge_Value::from)
+            .collect();
+
+        let func_type = unsafe { wasmedge::WasmEdge_VMGetFunctionType(self.ctx, raw_func_name) };
+        if func_type.is_null() {
+            return Err(ErrReport::default());
```

Comment:
May we use following ErrReport for this error?
```
ErrReport {
  code: None,
  message: "wasmedge get function type not found"
}
```

It is good if the code is `None`, the developer can know there is no error code from wasmedge VM, and the message describe something happened in the `-sys` layer.

Reply:
The `ErrReport`  field type is currently numeric and receives the error code from the c side. When the pointer is empty, there is actually no error code on the c side. Maybe I can leave the null pointer out of the error handling and just panic if there is a null pointer

Reply:
I think it is excellent to use `panic!` here if people calling the non-exist function. :+1: 

If the `panic!` is used, do we still need the `Default` trait for `ErrReport`?

It may be good to remove the `Default` implementation and let the `ErrReport` always be created from a `wasmedge::WasmEdge_Result`.